dartsim/src/FreeGroupFeatures_TEST.cc

```diff
@@ -0,0 +1,194 @@
+#include "ignition/physics/sdf/ConstructNestedModel.hh"
+#include "ignition/physics/sdf/ConstructWorld.hh"
+#include "test/Utils.hh"
+
+struct TestFeatureList : ignition::physics::FeatureList<
+    ignition::physics::sdf::ConstructSdfWorld,
+    ignition::physics::sdf::ConstructSdfModel,
+    ignition::physics::sdf::ConstructSdfNestedModel,
+    ignition::physics::GetWorldFromEngine,
+    ignition::physics::GetModelFromWorld,
+    ignition::physics::GetNestedModelFromModel,
+    ignition::physics::GetLinkFromModel,
+    ignition::physics::LinkFrameSemantics,
+    ignition::physics::FindFreeGroupFeature,
+    ignition::physics::SetFreeGroupWorldPose > { };
+
+using World = ignition::physics::World3d<TestFeatureList>;
+using WorldPtr = ignition::physics::World3dPtr<TestFeatureList>;
+using ModelPtr = ignition::physics::Model3dPtr<TestFeatureList>;
+using LinkPtr = ignition::physics::Link3dPtr<TestFeatureList>;
+
+/////////////////////////////////////////////////
+auto LoadEngine()
+{
+  ignition::plugin::Loader loader;
+  loader.LoadLib(dartsim_plugin_LIB);
+
+  ignition::plugin::PluginPtr dartsim =
+      loader.Instantiate("ignition::physics::dartsim::Plugin");
+
+  auto engine =
+      ignition::physics::RequestEngine3d<TestFeatureList>::From(dartsim);
+  return engine;
+}
+
+/////////////////////////////////////////////////
+WorldPtr LoadWorld(const std::string &_world)
+{
+  auto engine = LoadEngine();
+  EXPECT_NE(nullptr, engine);
+  if (nullptr == engine)
+  {
+    return nullptr;
+  }
+
+  sdf::Root root;
+  const sdf::Errors &errors = root.Load(_world);
+  EXPECT_EQ(0u, errors.size()) << errors;
+
+  EXPECT_EQ(1u, root.WorldCount());
+  const sdf::World *sdfWorld = root.WorldByIndex(0);
+  EXPECT_NE(nullptr, sdfWorld);
+
+  auto world = engine->ConstructWorld(*sdfWorld);
+  EXPECT_NE(nullptr, world);
+
+  return world;
+}
+
+ModelPtr GetModelFromAbsoluteName(const WorldPtr &_world,
+                                  const std::string &_absoluteName)
+{
+  std::vector<std::string> names =
+      ignition::common::split(_absoluteName, sdf::kSdfScopeDelimiter);
+  if (names.empty())
+  {
+    return nullptr;
+  }
+
+  auto currentModel = _world->GetModel(names.front());
+  for (std::size_t i = 1; i < names.size(); ++i)
+  {
+    if (nullptr == currentModel)
+      return nullptr;
+    currentModel = currentModel->GetNestedModel(names[i]);
+  }
+  return currentModel;
+}
+
+TEST(FreeGroupFeatures, NestedFreeGroup)
+{
+  WorldPtr world = LoadWorld(TEST_WORLD_DIR "world_with_nested_model.sdf");
+  ASSERT_NE(nullptr, world);
+
+  auto checkFreeGroupForModel = [&](const std::string &_modelName)
+  {
+    auto model = GetModelFromAbsoluteName(world, _modelName);
+    if (nullptr == model)
+      return testing::AssertionFailure() << "Model could not be found";
+    auto freeGroup = model->FindFreeGroup();
+    if (nullptr == freeGroup)
+      return testing::AssertionFailure() << "Freegroup could not be found";
+    if (nullptr == freeGroup)
```

Comment:
this logic statement is identical to the one before it and won't be reached, since the other would return first

Reply:
Thanks for catching that. I believe I meant to check if the Root (Canonical) Link is found. https://github.com/ignitionrobotics/ign-physics/pull/231/commits/91843ac3ccbf9a36605655ef732b1db44c769404